File: app/src/main/java/de/nils_beyer/android/Vertretungen/LoginActivity.java
package de.nils_beyer.android.Vertretungen;

import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.os.Vibrator;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;

import java.io.IOException;
import java.net.URL;

import javax.net.ssl.HttpsURLConnection;

import de.nils_beyer.android.Vertretungen.account.AccountSpinner;
import de.nils_beyer.android.Vertretungen.account.StudentAccount;
import de.nils_beyer.android.Vertretungen.download.StudentDownloadService;

public class LoginActivity extends AppCompatActivity {

    EditText username;
    EditText password;
    Button btn_login;
    TextView text_error;
    AccountSpinner accountSpinner;
    View contentView;
    ProgressBar progressbar;

    public boolean canClose;
    public static String KEY_REGISTERED = "REGISTERED_KEY_ID";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        // Set canClose:
        // If we have only unregistered Accounts
        // we cannot close the activity
        canClose = !AccountSpinner.hasOnlyUnregistered(getApplicationContext());

        if (Build.VERSION.SDK_INT >= 23)
            getWindow().setStatusBarColor(getResources().getColor(R.color.login_color, getTheme()));

        username = (EditText) findViewById(R.id.login_username);
        password = (EditText) findViewById(R.id.login_password);
        btn_login = (Button) findViewById(R.id.login_btn_login);
        text_error = (TextView) findViewById(R.id.login_error);
        contentView = findViewById(R.id.activity_login);
        accountSpinner = (AccountSpinner) findViewById(R.id.login_account_spinner);
        accountSpinner.setViewConfig(AccountSpinner.ViewConfig.SHOW_UNREGISTERED);
        progressbar = (ProgressBar) findViewById(R.id.progressBar);
        progressbar.setVisibility(View.GONE);


        btn_login.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                hideKeyboard();
                progressbar.setVisibility(View.VISIBLE);
                btn_login.setEnabled(false);
                checkPassword(new CheckPasswordCallback() {
                    @Override
                    public void onResult(boolean accepted) {
                        progressbar.setVisibility(View.GONE);
                        btn_login.setEnabled(true);
                        if (accepted) {
                            canClose = true;
                            finish();
                        } else {
                            text_error.setVisibility(View.VISIBLE);
                            ((Vibrator) getSystemService(VIBRATOR_SERVICE)).vibrate(100);
                        }
                    }

                    @Override
                    public void onError(Exception e) {
                        progressbar.setVisibility(View.GONE);
                        btn_login.setEnabled(true);
                        Snackbar.make(contentView, getString(R.string.io_error), Snackbar.LENGTH_LONG).show();
                    }
                });
            }
        });
    }

    @Override
    public void finish() {
        if (!canClose) {
            finishAffinity();
        } else {
            super.finish();
        }
    }


    protected void checkPassword(final CheckPasswordCallback checkPasswordCallback) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    boolean result = accountSpinner.tryRegister(username.getText().toString(), password.getText().toString());
                    if (result) {
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                checkPasswordCallback.onResult(true);

                            }
                        });
                    } else {
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                checkPasswordCallback.onError(new Exception());
                            }
                        });
                    }
                } catch (SecurityException e) {
                    // In case that the authorization
                    // credentials are wrong
                    // a security Exception is thrown by the
                    // DownloadService
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            checkPasswordCallback.onResult(false);
                        }
                    });
                } catch (final Exception e) {
                    // If there is an IOException
                    // the onError method will show a
                    // Snackbar with further information
                    // for the user.
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            checkPasswordCallback.onError(e);
                        }
                    });
                }
            }
        }).start();
    }

    private void hideKeyboard() {
        View view = this.getCurrentFocus();
        if (view != null) {
            InputMethodManager imm = (InputMethodManager)getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }

    private interface CheckPasswordCallback {
        void onResult(boolean accepted);
        void onError(Exception e);
    }
}

File: app/src/main/java/de/nils_beyer/android/Vertretungen/LoginActivity.kt
package de.nils_beyer.android.Vertretungen
import kotlinx.android.synthetic.main.activity_login.*


import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.VibrationEffect
import android.os.Vibrator
import com.google.android.material.snackbar.Snackbar
import androidx.appcompat.app.AppCompatActivity
import android.view.View
import android.view.inputmethod.InputMethodManager
import de.nils_beyer.android.Vertretungen.account.AccountSpinner
import de.nils_beyer.android.Vertretungen.download.UnAuthorizedException


class LoginActivity : AppCompatActivity() {
    var canClose: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)
        login_account_spinner.viewOrder = AccountSpinner.ViewOrder.INVERSE
        login_account_spinner.setViewConfig(AccountSpinner.ViewConfig.SHOW_UNREGISTERED)

        // Set canClose:
        // If we have only unregistered Accounts
        // we cannot close the activity
        canClose = !login_account_spinner.hasOnlyUnregistered(applicationContext)


        activity_login_text_version.text = String.format(getString(R.string.info_app_with_version), getAppVersion())

        if (Build.VERSION.SDK_INT >= 23) {
            window.statusBarColor = resources.getColor(R.color.login_color, theme)
        }

        // hide the progressbar
        progressBar.visibility = View.GONE


        // Setup Login Button
        login_btn_login.setOnClickListener({
            onLoginClicked()
        })
    }

    private fun onLoginClicked() {
        login_error.visibility = View.GONE
        hideKeyboard()

        progressBar.visibility = View.VISIBLE
        login_error.isEnabled = false

        // Try to login
        checkPassword({ accepted : Boolean ->
            // on result
            progressBar.visibility = View.GONE
            login_btn_login.isEnabled = true
            if (accepted) {
                canClose = true
                finish()
            } else {
                login_error.visibility = View.VISIBLE
                login_error.text = getString(R.string.login_error)
                val vibratorService = getSystemService(Context.VIBRATOR_SERVICE) as Vibrator

                // Use new VibrationEffect call on Android Oreo
                // if not available, use the deprecated vibrate method.
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    vibratorService.vibrate(VibrationEffect.createOneShot(
                            100, VibrationEffect.DEFAULT_AMPLITUDE
                    ))
                } else {
                    vibratorService.vibrate(100)
                }
            }
        }, {
            // on error
            progressBar.visibility = View.GONE
            login_btn_login.isEnabled = true

            login_error.visibility = View.VISIBLE
            login_error.text = it.message

            val vibratorService = getSystemService(Context.VIBRATOR_SERVICE) as Vibrator

            // Use new VibrationEffect call on Android Oreo
            // if not available, use the deprecated vibrate method.
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                vibratorService.vibrate(VibrationEffect.createOneShot(
                        100, VibrationEffect.DEFAULT_AMPLITUDE
                ))
            } else {
                vibratorService.vibrate(100)
            }

            Snackbar.make(activity_login, String.format(getString(R.string.activity_login_generic_error_template), it.message), Snackbar.LENGTH_LONG).show()
        })
    }

    override fun finish() {
        // When the User should not return to the MainActivity
        // because he is not logged in,
        // we close the App.
        if (!canClose) {
            finishAffinity()
        } else {
            super.finish()
        }
    }



    /**
     * Makes an Request to the Server to check Username/Password
     * @param checkPasswordCallback Callback for Success/Error/Failure
     */
    private fun checkPassword(callback: (Boolean)->(Unit), error : (Exception) -> Unit) {
        Thread(Runnable {
            try {
                login_account_spinner.getSelectedAccount().tryRegister(this, login_username.text.toString(), login_password.text.toString())
                runOnUiThread { callback(true) }
            } catch (e: UnAuthorizedException) {
                // In case that the authorization
                // credentials are wrong
                // a security Exception is thrown by the
                // DownloadService
                runOnUiThread { callback(false) }
            } catch (e: Exception) {
                // If there is an IOException
                // the onError method will show a
                // Snackbar with further information
                // for the user.
                runOnUiThread { error(e) }
            }
        }).start()
    }

    /**
     * Hides the On-board Keyboard
     */
    private fun hideKeyboard() {
        val view = this.currentFocus
        if (view != null) {
            val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        }
    }

    /**
     * Returns the app version name or null if packageManager throws an error
     */
    private fun getAppVersion() : String? {
        return try {
            val pInfo = packageManager.getPackageInfo(packageName, 0)
            pInfo.versionName
        } catch (e: PackageManager.NameNotFoundException) {
            null
        }
    }
}
